sentence.go

```diff
@@ -74,69 +78,125 @@ func (s BaseSentence) TalkerID() string {
+	return &SentenceParser{
+		parserLock:    &sync.RWMutex{},
+		customParsers: map[string]ParserFunc{},
+	}
+}
+
+// ParseBaseSentence parses BaseSentence from input
+func (p *SentenceParser) ParseBaseSentence(raw string) (BaseSentence, error) {
 	raw = strings.TrimSpace(raw)
-	tagBlockParts := strings.SplitN(raw, `\`, 3)
 
 	var (
 		tagBlock TagBlock
 		err      error
 	)
-	if len(tagBlockParts) == 3 {
-		tags := tagBlockParts[1]
-		raw = tagBlockParts[2]
-		tagBlock, err = parseTagBlock(tags)
+	if raw[0] == TagBlockSep {
+		// tag block is always at the start of line. Starts with `\` and ends with `\` and has valid sentence
+		// following or <CR><LF>
+		//
+		// Note: tag block group can span multiple lines but we only parse ones that have sentence
+		endOfTagBlock := strings.LastIndexByte(raw, TagBlockSep)
+		if endOfTagBlock <= 0 {
+			return BaseSentence{}, fmt.Errorf("nmea: sentence tag block is missing '\\' at the end")
+		}
+		tagBlock, err = parseTagBlock(raw[1:endOfTagBlock])
 		if err != nil {
 			return BaseSentence{}, err
 		}
+		if p.OnTagBlock != nil {
+			p.OnTagBlock(tagBlock)
+		}
+		raw = raw[endOfTagBlock+1:]
 	}
 
 	startIndex := strings.IndexAny(raw, SentenceStart+SentenceStartEncapsulated)
 	if startIndex != 0 {
 		return BaseSentence{}, fmt.Errorf("nmea: sentence does not start with a '$' or '!'")
 	}
-	sumSepIndex := strings.Index(raw, ChecksumSep)
-	if sumSepIndex == -1 {
-		return BaseSentence{}, fmt.Errorf("nmea: sentence does not contain checksum separator")
+	checksumSepIndex := strings.Index(raw, ChecksumSep)
+	fieldsRaw := raw[startIndex+1:]
+	checksumRaw := ""
+	if checksumSepIndex != -1 {
+		fieldsRaw = raw[startIndex+1 : checksumSepIndex]
+		checksumRaw = strings.ToUpper(raw[checksumSepIndex+1:])
 	}
-	var (
-		fieldsRaw   = raw[startIndex+1 : sumSepIndex]
-		fields      = strings.Split(fieldsRaw, FieldSep)
-		checksumRaw = strings.ToUpper(raw[sumSepIndex+1:])
-		checksum    = Checksum(fieldsRaw)
-	)
-	// Validate the checksum
-	if checksum != checksumRaw {
-		return BaseSentence{}, fmt.Errorf(
-			"nmea: sentence checksum mismatch [%s != %s]", checksum, checksumRaw)
+	if p.CheckCRC == nil {
+		if checksumSepIndex == -1 {
+			return BaseSentence{}, fmt.Errorf("nmea: sentence does not contain checksum separator")
+		}
+		if checksum := Checksum(fieldsRaw); checksum != checksumRaw {
+			return BaseSentence{}, fmt.Errorf("nmea: sentence checksum mismatch [%s != %s]", checksum, checksumRaw)
+		}
 	}
-	talker, typ := parsePrefix(fields[0])
-	return BaseSentence{
+
+	fields := strings.Split(fieldsRaw, FieldSep)
+	talker, typ, err := parsePrefix(fields[0])
+	if err != nil {
+		return BaseSentence{}, err
+	}
+	sentence := BaseSentence{
 		Talker:   talker,
 		Type:     typ,
 		Fields:   fields[1:],
 		Checksum: checksumRaw,
 		Raw:      raw,
 		TagBlock: tagBlock,
-	}, nil
+	}
+	if p.CheckCRC != nil {
+		if err := p.CheckCRC(checksumRaw, Checksum(fieldsRaw), sentence); err != nil {
```

Comment:
Why do we need to pass `checksumRaw` if it's already accessible on `BaseSentence.Checksum`?

Reply:
nope, `func(fieldsRaw string, sentence BaseSentence) error` is probably minimal. `fieldsRaw` so you can easily calculate own CRC without searching for `$`,`!` and `*` locations. `BaseSentence` so you can access everything else  - talkerid/type/whole raw line etc.